Update README for fluid telemetry package (#20549)
microsoft/FluidFramework: packages/framework/client-logger/fluid-telemetry/README.md

This package contains code enabling the production and consumption of typed telemetry for Fluid Framework applications. The typed telemetry from this package is used as the backbone for different Fluid Framework cloud offerings such as dashboards and alarms for Fluid applications. This package can also be used as a reference for customizing and creating your own telemetry solution if desired.

> At this time, the package enables collection of Fluid Container related telemetry. In the future, more areas would be added as needs evolve.

Reviewer: ```suggestion
At this time, the package enables collection of telemetry related to the Fluid Container only. In the future, more areas would be added as needs evolve.
```

# Getting Started

Let's walk through some examples for getting started with Fluid telemetry for containers using the `@fluidframework/fluid-telemetry` package.

## Use Case 1: Logging container telemetry to Azure App Insights and analyzing app data

### Prerequisite

> Before you can get telemetry sent to Azure App Insights, you'll need to create an Instance of App Insights on Azure. You'll then be able to leverage Azure App Insights integration and route your Fluid container application telemetry to App Insights. [Learn more about Azure App Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview). [Creating your App Insights instance](https://learn.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource?tabs=bicep).

Reviewer 2: ```suggestion
Before you can get telemetry sent to Azure App Insights, you'll need to create an Instance of App Insights on Azure. You'll then be able to integrate your Azure App Insights instance with your Fluid application and route your Fluid container application telemetry to App Insights. [Learn more about Azure App Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview). [Creating your App Insights instance](https://learn.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource?tabs=bicep).
```

nit on wording, no strong feelings on this though.

Once you setup your App Insights instance, you can proceed with next steps below to route the telemetry to App Insights.

### Step 1: Install new package dependencies

Install the Fluid Framework telemetry package and Azure App Insights package dependencies

-   Using NPM: `npm install @fluidframework/fluid-telemetry @microsoft/applicationinsights-web`

### Step 2: Setup telemetry collection

> Now, let's start the telemetry production by initializing our telemetry collection where we initialize our containers:

Reviewer 2: This was not part of your changes but on the example below on lines 32-33, we have two export lines from the same package that can be combined

// Done! Your container telemetry is now being created and sent to your Telemetry Consumer which will forward it to Azure App Insights.
```

That's it for now! If you've decided to use Azure App Insights, we have designed useful prebuilt queries that utilize the generated telemetry. You can try out these queries after running your application with above code for a little duration.

> You can find these queries [in the "Interpreting Telemetry Data" section below.](#telemetry_visualization).

Reviewer 2: ```suggestion
You can find these queries [in the "Interpreting Telemetry Data" section below.](#telemetry_visualization)
```
nit: extra period

## Use Case 2: Setup a custom telemetry consumer

To use a custom telemetry consumer and integrate with a cloud provider or custom dashboard solution, you need to implement the `ITelemetryConsumer` interface.

In this example, you'll walk through the basic setup process to start getting container telemetry to be produced and log it to the console.

### Step 1: Install new package dependencies

Install the Fluid Framework telemetry package dependency

-   Using NPM: `npm install @fluidframework/fluid-telemetry`

### Step 2: Define a custom telemetry consumer

> Here we'll create our own telemetry consumer which extends the `ITelemetryConsumer` interface. Let's look at an example that will simply log the telemetry to console.

Reviewer 2: This was not part of your changes but on line 113 the comment should be changed to `// 1: import our implementation of MySimpleTelemetryConsumer from step 2` (right now it says step 1)

// Done! Your container telemetry is now being created and sent to your Telemetry Consumer which will forward it to Azure App Insights.
```

That's it for now! If you've decided to use Azure App Insights, we have designed useful prebuilt queries that utilize the generated telemetry. You can try out these queries after running your application with above code for a little duration.

You can find these queries [here]().
// TODO: rohandubal:Fix me

## Use Case 2: Setup a custom telemetry consumer

> To use a custom telemetry consumer and integrate with a cloud provider or custom dashboard solution, you need to extend the ITelemetryConsumer package.

Reviewer: ```suggestion
To use a custom telemetry consumer and integrate with a cloud provider or custom dashboard solution, you need to implement the `ITelemetryConsumer` interface.
```

In this example, you'll walk through the basic setup process to start getting container telemetry to be produced and log it to the console.

### Step 1: Install new package dependencies

Install the Fluid Framework telemetry package dependency

-   Using NPM: `npm install @fluidframework/fluid-telemetry`

> ### Step 2: Setup a custom telemetry consumer

Reviewer: ```suggestion
### Step 2: Define a custom telemetry consumer
```

> Here we'll create our own telemetry consumer which extends the ITelemetryConsumer interface. Let's look at an example that will simply log the telemetry into console.

Reviewer: ```suggestion
Here we'll create our own telemetry consumer which extends the `ITelemetryConsumer` interface. Let's look at an example that will simply log the telemetry to console.
```

That's it for now! If you've decided to use Azure App Insights, we have designed useful prebuilt queries that utilize the generated telemetry. You can try out these queries after running your application with above code for a little duration.

> You can find these queries [here](#telemetry_visualization).

Reviewer: Nit, I was told once that for screen readers it is useful to have more descriptive text in links.
 ```suggestion
You can find these queries in [the "Interpreting telemetry data" section](#telemetry_visualization) below.
```

}
```

> ### Step 3: Setup the custom telemetry consumer

Reviewer: ```suggestion
### Step 3: Setup telemetry collection
```

> Note: All telemetry being visualized below is generated from clients without any intervention from the server. The accuracy might be impacted due to inherent nature of client telemetry and data being lost due to faulty clients or lack of connectivity before the telemetry is fully pushed out. We recommend not using this telemetry for business metrics, but rather use it for operational metrics and diagnosis of issues.

### Accessing App Insights Portal

Before we can query, we must first navigate to your Azure App Insights telemetry page. To do this, go to your Azure App Insights Instance and click on the “Logs” tab under Monitoring.

![Logs on App Insights Portal](readme_images/telemetry_1.png)

Now, close out the “Queries” pane if it showed up for you and you will be in the view where we can execute our queries. Note that if you are using the Fluid Azure App Insights logger, your telemetry data will be available in the “customEvents” table. 

### Queries and results

1. Session information

> The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the container being interacted with, the number of collaborators and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; see the intro paragraph for more information on sessions.

Reviewer 2: We might want to adjust this wording to number of containers to avoid confusion. I have a new image if you choose to update the query in my suggestion below. We can also just clarify what we mean instead and keep the image/query the same if you like.
```suggestion
The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the container being interacted with, the number of containers and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; see the intro paragraph for more information on sessions.
```
![image](https://github.com/microsoft/FluidFramework/assets/105244057/1b243fa8-0f47-4fdc-a7ed-7ddfebb4317a)

```sql
let sessionGap = 5m; 
let sessionGapSeconds = toint(sessionGap / 1s); 
customEvents 
| extend containerId = tostring(customDimensions.containerId), containerInstanceId = tostring(customDimensions.containerInstanceId ) 
| where name startswith "fluidframework.container"
| extend containerIdTimestamp = strcat(containerId, timestamp) 
| sort by containerIdTimestamp asc 
| extend prevTimestamp = prev(timestamp), prevContainerId = prev(containerId) 
| extend Diff = datetime_diff("second", timestamp, prevTimestamp) 
| extend IsNewPeriod = iif(prevContainerId != containerId or Diff > (sessionGapSeconds) or isnull(Diff), 1, 0) 
| extend SessionId = row_cumsum(IsNewPeriod) 

> | summarize NumCollaborators = dcount(containerInstanceId), StartTime = min(timestamp), EndTime = max(timestamp) by SessionId, containerId 
> | extend PeriodDurationInMinutes = datetime_diff("minute", EndTime, StartTime) 
> | project SessionId, containerId , NumCollaborators, PeriodDurationInMinutes, StartTime, EndTime 

Reviewer 2: ```suggestion
| summarize NumContainers = dcount(containerInstanceId), StartTime = min(timestamp), EndTime = max(timestamp) by SessionId, containerId 
| extend PeriodDurationInMinutes = datetime_diff("minute", EndTime, StartTime) 
| project SessionId, containerId , NumContainers, PeriodDurationInMinutes, StartTime, EndTime 
```

We might want to rename this from Collaborators to NumContainers, or just make it clear that its not like unique people. (FYI if you make this update you'll need an updated image. I can provide if needed)

Reviewer 2: ![image](https://github.com/microsoft/FluidFramework/assets/105244057/1c5dfce5-e8e7-46e0-87e5-1b5731385aa6)

// Done! Your container telemetry is now being created and sent to your Telemetry Consumer
```

You can now run the app and see the telemetry being printed on your console.

> ## <a name="telemetry_visualization">Interpreting telemetry data</a>

Reviewer: Oh, interesting, is this a standard thing with markdown? I would have used `[link text](#interpreting-telemetry-data)` above but fixing an id seems way better.

Author: yup! I learnt it today also. Was not aware previously; this comes in handy.

Going into more detail, A “Session” or “Collaborative Session” is defined as a period in which we see a continuous stream of container telemetry being emitted with a unique container Id. For a given set of telemetry with the same container Id, if this stream of telemetry stops for longer than a specified period of time then we consider that the end of the session. Each session is differentiated from each other because it has no telemetry (no user activity) emitted for a defined amount of time prior to the start and after the end of emitted telemetry. For each of these queries you can also adjust the time gap that defines a session based on your preference. 

> Note: All telemetry being visualized below is generated from clients without any intervention from the server. The accuracy might be impacted due to inherent nature of client telemetry and data being lost due to faulty clients or lack of connectivity before the telemetry is fully pushed out. We recommend not using this telemetry for business metrics, but rather use it for operational metrics and diagnosis of issues.

### Accessing App Insights Portal

> Before we can query, we must first navigate to your Azure App Insights telemetry page. To do this, go to your Azure App Insights Instance and Click on the “Logs” tab under Monitoring

Reviewer: ```suggestion
Before we can query, we must first navigate to your Azure App Insights telemetry page. To do this, go to your Azure App Insights Instance and click on the “Logs” tab under Monitoring.
```

![Logs on App Insights Portal](readme_images/telemetry_1.png)

> Now, close out the “Queries” pane if it showed up for you and you will be in the view where we can now execute our queries. Note that if you are using the Fluid Azure App Insights logger, your telemetry data will be available in the “customEvents” table. 

Reviewer: Should this make reference to this package specifically instead of the "Fluid Azure App Insights Logger"? Or is it intentionally referring to the _other_ package, that pushes the internal telemetry?
```suggestion
Now, close out the “Queries” pane if it showed up for you and you will be in the view where we can execute our queries. Note that if you are using the Fluid Azure App Insights logger, your telemetry data will be available in the “customEvents” table. 
```

### Queries and results

1. Session information

> The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the document being interacted with, the number of collaborators and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; See the intro paragraph for more information on sessions.

Reviewer: ```suggestion
The following query provides a table of data that can give you a quick overview of information about sessions for your application. It includes the Id of the container being interacted with, the number of collaborators and the length of each session. Note that query provides session id’s but these values do not actually exist in the telemetry, it is a concept we have derived from the data; see the intro paragraph for more information on sessions.
```

```sql
let sessionGap = 5m; 
let sessionGapSeconds = toint(sessionGap / 1s); 
customEvents 
| extend containerId = tostring(customDimensions.containerId), containerInstanceId = tostring(customDimensions.containerInstanceId ) 
| where name startswith "fluidframework.container"
| extend containerIdTimestamp = strcat(containerId, timestamp) 
| sort by containerIdTimestamp asc 
| extend prevTimestamp = prev(timestamp), prevContainerId = prev(containerId) 
| extend Diff = datetime_diff("second", timestamp, prevTimestamp) 
| extend IsNewPeriod = iif(prevContainerId != containerId or Diff > (sessionGapSeconds) or isnull(Diff), 1, 0) 
| extend SessionId = row_cumsum(IsNewPeriod) 
| summarize NumCollaborators = dcount(containerInstanceId), StartTime = min(timestamp), EndTime = max(timestamp) by SessionId, containerId 
| extend PeriodDurationInMinutes = datetime_diff("minute", EndTime, StartTime) 
| project SessionId, containerId , NumCollaborators, PeriodDurationInMinutes, StartTime, EndTime 
```
![Query result](readme_images/telemetry_2.png)

2. Total number of sessions over time period

> The following query provides the average number of total sessions occurring over 10-minute intervals. The query logic graphs the total number of sessions occurring over 10-minute data points denoted by the variable named summedDataPointInterval. 

Reviewer: ```suggestion
The following query provides the total number of sessions seen over 10-minute intervals. The query logic graphs the total number of sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. 
```

let minTimestamp = toscalar(customEvents | where name startswith "fluidframework.container" | summarize min(timestamp)); 
let maxTimestamp = toscalar(customEvents | where name startswith "fluidframework.container" | summarize max(timestamp)); 
customEvents 
| extend containerId = tostring(customDimensions.containerId) 
| where name startswith "fluidframework.container" 
| extend docIdTimestamp = strcat(containerId, timestamp) 
| sort by docIdTimestamp asc 
| extend prevTimestamp = prev(timestamp), prevContainerId = prev(containerId) 
| extend Diff = datetime_diff("second", timestamp, prevTimestamp) 
| extend IsNewPeriod = iif(prevContainerId != containerId or Diff > (sessionGapSeconds) or isnull(Diff), 1, 0) 
| extend SessionId = row_cumsum(IsNewPeriod) 
| summarize by SessionId, containerId, bin(timestamp, summedDataPointInterval) 
| make-series sumSessions = dcount(SessionId) on timestamp from minTimestamp to maxTimestamp step averagedTimeInterval 
| render timechart with (title = "Total Number of Sessions Occurring Over 1 Hour Intervals") 
```
![Total sessions over time period](readme_images/telemetry_3.png)

3. Average number of sessions over time period

> The following query provides the average number of total sessions occurring over 1-hour intervals. The query logic first sums up the total number of sessions occurring over 10-minute data points denoted by the variable named summedDataPointInterval. It then averages these data points over 1-hour intervals denoted by the averagedTimeInterval variable.

Reviewer: ```suggestion
The following query provides the average number of sessions occurring over 1-hour intervals. The query logic first sums up the total number of sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. It then averages these data points over 1-hour intervals denoted by the `averagedTimeInterval` variable.
```

let maxTimestamp = toscalar(customEvents | where name startswith "fluidframework.container" | summarize max(timestamp)); 
customEvents 
| extend containerId = tostring(customDimensions.containerId)
| where name startswith "fluidframework.container"
| extend containerIdTimestamp = strcat(containerId, timestamp) 
| sort by containerIdTimestamp asc 
| extend prevTimestamp = prev(timestamp), prevContainerId = prev(containerId) 
| extend Diff = datetime_diff("second", timestamp, prevTimestamp) 
| extend IsNewPeriod = iif(prevContainerId != containerId or Diff > (sessionGapSeconds) or isnull(Diff), 1, 0) 
| extend SessionId = row_cumsum(IsNewPeriod) 
| summarize sumSessions = dcount(SessionId) by bin(timestamp, summedDataPointInterval)
| make-series avgSessions = avg(sumSessions) on timestamp from minTimestamp to maxTimestamp step averagedTimeInterval 
| render timechart with (title = "Average Number of Sessions Occurring Over 1 Hour Intervals")
```

![Average number of sessions over time period](readme_images/telemetry_4.png)

4. Average number of containers per session over a time period

> The following query will provide you with the approximate average number of containers per sessions over 1 hour time intervals. The query logic first sums up the total number of collaborators per sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. It then averages these datapoints over 1-hour intervals denoted by the `averagedTimeInterval` variable.

Reviewer: ```suggestion
The following query will provide you with the approximate average number of containers per session over 1 hour time intervals. The query logic first sums up the total number of containers per sessions occurring over 10-minute data points denoted by the variable named `summedDataPointInterval`. It then averages these datapoints over 1-hour intervals denoted by the `averagedTimeInterval` variable.
```

customEvents 
| extend containerId = tostring(customDimensions.containerId), containerInstanceId = tostring(customDimensions.containerInstanceId) 
| where name startswith "fluidframework.container"
| extend docIdTimestamp = strcat(containerId, timestamp) 
| sort by docIdTimestamp asc 
| extend prevTimestamp = prev(timestamp), prevContainerId = prev(containerId) 
| extend Diff = datetime_diff("second", timestamp, prevTimestamp) 
| extend IsNewPeriod = iif(prevContainerId != containerId or Diff > (sessionGapSeconds) or isnull(Diff), 1, 0) 
| extend SessionId = row_cumsum(IsNewPeriod) 
| summarize sumCollaborators = dcount(containerInstanceId) by SessionId, containerId, bin(timestamp, summedDataPointInterval)
| make-series avgCollaborators = avg(sumCollaborators) on timestamp from minTimestamp to maxTimestamp step averagedTimeInterval
| render timechart with (title = "Approximate Average Number Of Container Per Session Over 1 Hour Intervals")
```

![Average number of containers per session over time period](readme_images/telemetry_5.png)

5. Length of Individual Sessions in Minutes

> This query provides you with the length of time of individual sessions, graphed by putting them into time bins increasing in 2.5 minute increments. The query logic calculates the length of time of each session that occurs and then sorts them into time bins going up in 2.5 minute increments denoted by the `let sessionLengthMinuteBins = 2.5;` variable. To adjust the 2.5 minute time bins, simply modify the `sessionLengthMinuteBins` variable, note that this number represents minutes.

Reviewer: ```suggestion
This query provides you with the length of time of individual sessions, graphed by putting them into time bins increasing in 2.5 minute increments. The query logic calculates the length of time of each session that occurs and then sorts them into time bins going up in 2.5 minute increments denoted by the `let sessionLengthMinuteBins = 2.5;` variable. To adjust the 2.5 minute time bins, simply modify the `sessionLengthMinuteBins` variable. Note that this number represents minutes.
```

| summarize StartTime = min(timestamp), EndTime = max(timestamp) by SessionId, containerId 
| extend SessionDurationInMinutes = datetime_diff("minute", EndTime, StartTime)
| make-series numSessions = dcount(SessionId) default=0 on sessionLengthBin = bin(SessionDurationInMinutes, sessionLengthMinuteBins) step sessionLengthMinuteBins
| mvexpand sessionLengthBin, numSessions
| extend numSessionsLong = tolong(numSessions), sessionLengthBinDouble = todouble(sessionLengthBin)
| project LengthOfSession = sessionLengthBinDouble, NumberOfSessions = numSessionsLong
| sort by LengthOfSession asc
| render columnchart with (title = "Length Of Sessions Separated Into 2.5 Minute Time Bins") 

```

![Length of Individual Sessions in Minutes](readme_images/telemetry_6.png)

#### General Query Adjustments

1. Adjusting the date span of the query 

To adjust the time span of this query, simply use the Time Range dropdown provided by azure. You do not need to modify the query directly. By default, these queries will query against all logs you have available.

![Adjusting date](readme_images/telemetry_time_period.png)

2. Adjusting the gap of time that defines a session

> By default, we identify each session has a period of 5 minutes of inactivity before and after. However, you may want to adjust this time period. To do so, modify the value of the variable named sessionGap to be your desired time length.

Reviewer: ```suggestion
By default, we identify each session has a period of 5 minutes of inactivity before and after. However, you may want to adjust this time period. To do so, modify the value of the variable named `sessionGap` to be your desired time length.
```
